src/Projection/ProcessEvents.php

```diff
@@ -0,0 +1,208 @@
+                \usleep($this->sleep);
+                $this->updateLock();
+            } else {
+                $this->persist();
+            }
+
+            $this->eventCounter = 0;
+
+            if ($this->triggerPcntlSignalDispatch) {
+                \pcntl_signal_dispatch();
+            }
+
+            switch ($this->fetchRemoteStatus()) {
+                case ProjectionStatus::STOPPING():
+                    $this->stop();
+                    break;
+                case ProjectionStatus::DELETING():
+                    $this->delete(false);
+                    break;
+                case ProjectionStatus::DELETING_INCL_EMITTED_EVENTS():
+                    $this->delete(true);
+                    break;
+                case ProjectionStatus::RESETTING():
+                    $this->reset();
+                    break;
+                default:
+                    break;
+            }
+
+            if (! $this->isStopped && $this->status === ProjectionStatus::RUNNING()) {
+                for ($changeStream->rewind(); true; $changeStream->next()) {
+                    if (! $changeStream->valid()) {
+                        \usleep($this->sleep);
+                        $this->updateLock();
+                        if ($this->triggerPcntlSignalDispatch) {
+                            \pcntl_signal_dispatch();
+                        }
+                        switch ($this->fetchRemoteStatus()) {
+                            case ProjectionStatus::STOPPING():
+                                $this->stop();
+                                break;
+                            case ProjectionStatus::DELETING():
+                                $this->delete(false);
+                                break;
+                            case ProjectionStatus::DELETING_INCL_EMITTED_EVENTS():
+                                $this->delete(true);
+                                break;
+                            case ProjectionStatus::RESETTING():
+                                $this->reset();
+                                break;
+                            default:
+                                break;
+                        }
+
+                        if ($this->isStopped || ! $keepRunning) {
+                            break;
+                        }
+                        continue;
+                    }
+
+                    $event = $changeStream->current();
+
+                    if ($event['operationType'] === 'invalidate') {
+                        break;
+                    }
+
+                    $streamName = $collectionNames[$event['ns']['coll']];
+                    $eventTimestamp = $event['clusterTime']->getTimestamp();
+
+                    // event already processed
+                    if ($eventTimestamp <= $streamTimestampsStart[$streamName]
+                        && $event['fullDocument']['_id'] <= $this->streamPositions[$streamName]
+                    ) {
```

Comment:
Already processed event is hard to test and I'm not sure if detects all cases.

Reply:
maybe it helps to draw some scenarios using draw.io?